### ISSUE TITLE
option changed at every puppet run?

### ISSUE THREAD

**Issue by gsgatlin** [mosen/puppet-cups#63]
I have a definition for a printer that looks like this:

printer { 'wolfprint-blackandwhite' :
    ensure  => present,
    uri     => 'ipps://print.ncsu.edu/printers/WolfPrint-BlackAndWhite',
    model   => 'drv:///sample.drv/generic.ppd',
    options => {
    'Option1' => 'True',
    },
    }

Every time puppet runs I get new output:

Notice: /Stage[main]/Itecsprinters::Pa_5/Printer[wolfprint-blackandwhite]/options: options changed {} to '{"Option1"=>"True"}'

Thus it appears this module is not idempotent? Should I be worried that it keeps changing options at every new puppet run? Thanks for any ideas you may have about this.

**Comment by mosen** [comment]
Chances are that the option is not actually being changed, therefore it tries to run every single time. 

You could try changing a standard option like ColorModel to Gray, or MediaType to envelope. There are still a few problems with changing options after the printer has been created though. Sometimes the PPD options are only set upon creation.

If you just want to test the module, you can just delete the options altogether!

**Comment by gsgatlin** [comment]
Thanks. I tried removing option1 and did a re-install to make sure everything was fresh and I can still print fine to those printers without 'Option1' => 'True', so I'll just leave that off. Thanks a lot.

**Comment by gsgatlin** [comment]
Ok. It is working.

printer { 'wolfprint-blackandwhite' :
    ensure  => present,
    uri     => 'ipps://print.ncsu.edu/printers/WolfPrint-BlackAndWhite',
    model   => 'drv:///sample.drv/generic.ppd',
    options => {
    'Option1' => 'True',
    },
    }

    printer { 'wolfprint-color' :
    ensure  => present,
    uri     => 'ipps://print.ncsu.edu/printers/WolfPrint-Color',
    model   => 'drv:///sample.drv/generic.ppd',
    }


grep DefaultOption1 /etc/cups/ppd/wolfprint*.ppd
/etc/cups/ppd/wolfprint-blackandwhite.ppd:*DefaultOption1: True
/etc/cups/ppd/wolfprint-color.ppd:*DefaultOption1: False


So, is there anything that can be done about the output at every run? Or should it be ignored?

Thanks so much.

**Comment by ih84ds** [comment]
I'm running into this issue as well, but only for certain options. Behavior seems to be this:

* All specified options are being set properly
* Most options do not cause a "change" on every puppet run
* A few options do force a "change" on every puppet run, e.g. InputSlot and PageSize

I noticed that the "few" options that do cause the constant changes don't show up in the "from" list, only the "to" list.

**Comment by ih84ds** [comment]
Scratch that, partially solved. InputSlot and PageSize should be set with input_tray and page_size and not from within ppd_options.

There is still one vendor option I am having issues with, and I see the problem. It does not match the \w+ regex because the value has a "-" in it. Will open a pull request with a patch for this.

**Comment by mosen** [comment]
Thanks I've merged that. I haven't been working on this lately but happy to take another look at the issue.

**Comment by ncsutmf** [comment]
Turns out 'Option1' is a vendor specific option for the generic postscript driver.  I think at the time we first started using it with this module, ppd_options was not an available parameter so making the connection between the issue and that particular change was not immediately obvious.

So the following code works and makes puppet logs a lot quieter:

``` puppet
printer { 'QUEUENAME' :
  ensure      => present,
  uri         => 'ipps://CUPS_SERVERNAME/printers/QUEUENAME',
  model       => 'drv:///sample.drv/generic.ppd',
  ppd_options => {
    'Option1' => 'True',
  },
}
```

I believe this issue can now be closed.
